homework/index.js

```diff
@@ -1,46 +1,169 @@
 'use strict';
 
 {
-  function fetchJSON(url, cb) {
-    const xhr = new XMLHttpRequest();
-    xhr.open('GET', url);
-    xhr.responseType = 'json';
-    xhr.onload = () => {
-      if (xhr.status >= 200 && xhr.status <= 299) {
-        cb(null, xhr.response);
-      } else {
-        cb(new Error(`Network error: ${xhr.status} - ${xhr.statusText}`));
-      }
-    };
-    xhr.onerror = () => cb(new Error('Network request failed'));
-    xhr.send();
-  }
-
   function createAndAppend(name, parent, options = {}) {
     const elem = document.createElement(name);
     parent.appendChild(elem);
     Object.keys(options).forEach(key => {
       const value = options[key];
       if (key === 'text') {
         elem.textContent = value;
+      } else if (key === 'html') {
+        elem.innerHTML = value;
       } else {
         elem.setAttribute(key, value);
       }
     });
     return elem;
   }
 
-  function main(url) {
-    fetchJSON(url, (err, repositories) => {
-      const root = document.getElementById('root');
-      if (err) {
-        createAndAppend('div', root, { text: err.message, class: 'alert-error' });
-        return;
-      }
-      createAndAppend('pre', root, { text: JSON.stringify(repositories, null, 2) });
+  function addRow(tbody, label, value) {
+    const tr = createAndAppend('tr', tbody);
+    createAndAppend('td', tr, { text: `${label}:` });
+    createAndAppend('td', tr, { text: value });
+    return tr;
+  }
+
+  function renderError(error, container) {
+    createAndAppend('div', container, { text: error.message, class: 'alert-error' });
+  }
+
+  async function bringData(url) {
+    const response = await fetch(url);
+    if (!response.ok) {
+      throw Error(`Network error: ${response.status} - ${response.statusText}`);
+    } else if (response.statusText !== 'OK') {
+      return null;
+    } else {
+      return response.json();
+    }
+  }
+
+  function renderContributions(contributors, infoRight) {
+    if (contributors === null) {
+      createAndAppend('p', infoRight, {
+        class: 'no-content',
+        html: 'This repository is empty.<br /> It has No content. <br /> It has No contributions.',
```

Comment:
when I get `null` from the previous comment line 36, that means this repository is empty. It has nothing but just a name, so no content. And sure  no contributors because no one writes anything there yet . 
There is another situation: repository with content but no contributors, that is something else.

Reply:
in the case (repository with content but no contributors) i get the response fro the server as empty array. 
But in the case of empty repository I got nothing, I got just an error because I am asking for something not existed yet.

Reply:
Your code that returns `null` in line 36 is incorrect as I tried to explain. That function is not specific to GitHub and returning `null` has no relation at all with GitHub.

A repository 'with just a name' is not necessarily empty. An 'empty' repository is one that contains no files at all, but that is not something your code checks.

Reply:
So I have to check that later again, but just to make my point clear and here is an empty repository from GitHub:
https://github.com/HackYourFuture/hyfer-infra

the name of the repository is  : hyfer-infra
you can test it on my homework: https://adham-elias-botours.github.io/my-js3/week03/
thanks in advanced.

Reply:
Okay, I see what you mean. First of all, I commend you for testing your code with many different repositories and finding this one.

But the request you make that results in the status text not being 'OK' is for the contributors, not for the repo. The only reasonable conclusion you can draw is that there are no contributors. Does that always mean that the repo is empty? Perhaps, but unless we can find that documented somewhere, we should not draw that conclusion.

Instead of checking for the word 'OK' it would safer to check the status code `!== 200'.

Reply:
thanks, But in the code there is another  scenario for non empty repository but has no contributions . 
this repo: https://github.com/HackYourFuture/DataStructures
Repo name is: DataStructures
you may would like to test it.. 
I am going to make a foto then the idee will be 100% clear,

Reply:
<img width="960" alt="empty" src="https://user-images.githubusercontent.com/50738429/62796954-c3ac5a00-bada-11e9-8669-16e53fb58c45.png">
<img width="976" alt="zero" src="https://user-images.githubusercontent.com/50738429/62796964-cad36800-bada-11e9-9956-4024b88251fd.png">

Reply:
I have admin rights on the HYF GitHub organization. Below is what I see. But, I appreciate all the research you have done and finding some exceptional repos. I will be helpful for me to know about it for the next classes.

![image](https://user-images.githubusercontent.com/2788771/62797275-97dda400-badb-11e9-8c14-f76fe4b431e7.png)

Reply:
Thanks for your time.  
I am struggling in handling errors, but after a time I may find the good way. 
Best regards